### DISCUSSION TITLE
Performance degradation between 3.7.12 and 4.6.3

### DISCUSSION THREAD

**Discussion by heitbaum** [ccache/ccache#1160] [Support]
Hi All,

we are preparing the update of the ccache in LibreELEC from 3.7.12 to 4.6.3 in https://github.com/LibreELEC/LibreELEC.tv/pull/6887 but are experiencing a performance degradation of ~4%. 

We have done a controlled “from clean” test in:
- https://github.com/LibreELEC/LibreELEC.tv/pull/6887#issuecomment-1249231988
- https://github.com/LibreELEC/LibreELEC.tv/pull/6887#issuecomment-1249252094

Is this degradation expected?
Are we doing something wrong?

We believe we have ruled out “compression” being the reason.

**Comment by afbjorklund** [comment]
There should be some tracing support, if you want to see _which_ parts are taking the time:

https://github.com/ccache/ccache/blob/master/doc/DEVELOPER.md

* https://github.com/ccache/ccache/pull/280

The main other difference (besides compression*) would be checksum, and the C++ rewrite ?

----

\* if you are concerned about compression performance in ccache 3.x I would suggest using
either gzip -1 (--fast) or lz4 - or stick with disabling compression and go for a bigger cache size

https://facebook.github.io/zstd/#benchmarks

with the right compile settings, the BLAKE3 checksum in ccache 4.x should be _faster_ than MD4 
but it cannot be turned off, so you would have to benchmark it separately - if it concerns you

https://github.com/BLAKE3-team/BLAKE3#readme

**Comment by afbjorklund** [comment]
One thing to watch out for, is that cmake compiles everything **without optimization** as the default!

So you need to make sure to switch to "RelWithDebInfo" to get the usual `-O2 -g` compile flags...

https://cmake.org/cmake/help/v3.16/variable/CMAKE_BUILD_TYPE.html

> Possible values are empty, Debug, Release, RelWithDebInfo, MinSizeRel, …

**Comment by jrosdahl** [comment]
Hi,

> Is this degradation expected?

I'm not sure what the test you performed actually did so it's hard for me to say.

Generally speaking I wouldn't say that you should _expect_ degradation, but I would also not be surprised if there is some scenario where 4.6.3 is slower than 3.7.12. However, if ccache 4.6.3 is 4% slower than 3.7.12 for a cache miss build then it sounds like a lot and would indeed be surprising. The 4.x versions have mostly been focused on features, not performance, but I can't think of any changes that would motivate a major slowdown in cache miss cases, and I also can't reproduce any such slowdown. For cache hit scenarios, I can reproduce that 4.6.3 is slower than 3.7.12 for large object files.

Incidentally, I have made some substantial rewrites of storage handling on master (will be part of ccache 4.7) and it is now faster than 3.7.12 in all my test cases.

Another thing to note is that you could try enabling [the depend mode](https://ccache.dev/manual/4.6.3.html#_the_depend_mode), which will make the ccache overhead for cache misses much smaller. There are some trade-offs that can be read in the documentation linked above.
